docs: add documentation for using lambda functions (#7092)
confluentinc/ksql: docs/how-to-guides/use-lambda-functions.md

> # Use lambda functions

Reviewer: Add YAML front matter (just crib from another md file).

Author: Where would I find it? It doesn't look like any of the other how-to guides have it unless my IDE is just hiding it

Reviewer: Heh, well look at that! Every page should have yaml -- go ahead and merge and I'll add to  the frontmatter to all of these at once.

## Context

> There doesn't exist a built-in UDF that suits your needs and you're unable to implement and deploy a UDF.

Reviewer: ```suggestion
There's no built-in UDF that suits your needs, and you can't implement and deploy a UDF.
```

> To get around this limitation, you use ksqlDB implemented lambda functions.

Reviewer: ```suggestion
To get around this limitation, use ksqlDB lambda functions.
```

## In action
```sql
CREATE STREAM stream1 (
  id INT,
  lambda_map MAP<STRING, INTEGER>
) WITH (
  kafka_topic = 'stream1',
  partitions = 1,
  value_format = 'avro'
);

CREATE STREAM output AS
  SELECT id, 
  TRANSFORM(lambda_map, (k, v) => UCASE(k), (k, v) => v + 5) 
  FROM stream1
  EMIT CHANGES;
```

## Syntax

> The arguments for the lambda function are separated from the body of the lambda with a `=>`.

Reviewer: ```suggestion
The arguments for the lambda function are separated from the body of the lambda with the lambda operator, `=>`.
```

> When there are 2 or more arguments the arguments must be enclosed with parentheses. Parentheses are optional for lambda functions with 1 argument.

Reviewer: ```suggestion
When there are two or more arguments, you must enclose the arguments with parentheses. Parentheses are optional for lambda functions with one argument.
```

> Currently, ksqlDB only supports up to 3 arguments in a single lambda function.

Reviewer: ```suggestion
Currently, ksqlDB supports up to three arguments in a single lambda function.
```

```sql
x => x + 5

(x,y) => x - y

(x,y,z) => z AND x OR y
```

## Invocation UDFs

> Lambda functions can only be used inside designated invocation functions. Invocations available currently are:

Reviewer: ```suggestion
Lambda functions must be used inside designated invocation functions. These are the available Invocations:
```

- [TRANSFORM](/developer-guide/ksqldb-reference/scalar-functions#TRANSFORM)
- [REDUCE](/developer-guide/ksqldb-reference/scalar-functions#REDUCE)
- [FILTER](/developer-guide/ksqldb-reference/scalar-functions#FILTER)

> ## Create a Lambda Compatible Stream

Reviewer: ```suggestion
## Create a lambda-compatible stream
```

> Invocation functions require either a map or array input. Here, we create a stream

Reviewer: ```suggestion
Invocation functions require either a map or array input. The following example creates a stream
```

> with a column type `MAP<STRING, INTEGER>`.

Reviewer: ```suggestion
with a column type of `MAP<STRING, INTEGER>`.
```

```sql
CREATE STREAM stream1 (
  id INT,
  lambda_map MAP<STRING, INTEGER>
) WITH (
  kafka_topic = 'stream1',
  partitions = 1,
  value_format = 'avro'
);
```

## Apply a lambda invocation function

> A lambda invocation function is a [scalar UDF](/developer-guide/ksqldb-reference/scalar-functions) and is used as such.

Reviewer: ```suggestion
A lambda invocation function is a [scalar UDF](/developer-guide/ksqldb-reference/scalar-functions), and you use it the same way as other scalar functions.
```

> This lambda function transforms both the key and value of a map and produces a new map. The key is transformed

Reviewer: ```suggestion
The following example lambda function transforms both the key and value of a map and produces a new map. The key is transformed
```

into an uppercase string using a built in UDF, and the value is transformed through addition. The order of the variables

> matters, the first item in the arguments list, here `k`, is treated as the key and the second, here `v` is treated

Reviewer: ```suggestion
is important: the first item in the arguments list, named `k` this example, is treated as the key, and the second, named `v` this example, is treated
```

as the value. Pay attention to this if your map has different types. Note that `transform` on
a map requires two lambda functions, while `transform` on an array requires one.
```sql
CREATE STREAM output AS
  SELECT id, 
  TRANSFORM(lambda_map, (k, v) => UCASE(k), (k, v) => v + 5) 
  FROM stream1;
```

Insert some values into `stream1`.
```sql
INSERT INTO stream1 (
  id, lambda_map
) VALUES (
  3, MAP("hello":= 15, "goodbye":= -5)
);
```

Query the output.
```sql
SELECT * FROM output AS final_output;
```

> You should see something similar to:

Reviewer: ```suggestion
Your output should resemble:
```

```sql
+------------------------------+------------------------------+
|id                            |final_output                  |
+------------------------------+------------------------------+
|3                             |{HELLO: 20}                   |
|4                             |{GOODBYE: 0}                  |                           
```

## Use a reduce lambda invocation function

> Create a stream with a column type `ARRAY<INTEGER>` and apply the `reduce` lambda 

Reviewer: ```suggestion
The following example creates a stream with a column type `ARRAY<INTEGER>` and applies the `reduce` lambda 
```

```
Insert some values into `stream1`.
```sql
INSERT INTO stream1 (
  id, lambda_arr
) VALUES (
  1, ARRAY(2, 3, 4, 5)
);
```

Query the output.
```sql
SELECT * FROM output AS final_output;
```

You should see something similar to:
```sql
+------------------------------+------------------------------+
|id                            |final_output                  |
+------------------------------+------------------------------+
|1                             |{output:3}                    |  
```

## Use a filter lambda invocation function
Create a stream with a column type `MAP<STRING, INTEGER>`and apply the `filter` lambda 
invocation function. 
```sql
CREATE STREAM stream1 (
  id INT,
  lambda_map MAP<STRING, INTEGER>
) WITH (
  kafka_topic = 'stream1',
  partitions = 1,
  value_format = 'avro'
);

CREATE STREAM output AS
  SELECT id, 
  FILTER(lambda_map, (k, v) => instr(k, 'name') > 0 AND v != 0) 
  FROM stream1
  EMIT CHANGES;
```
Insert some values into `stream1`.
```sql
INSERT INTO stream1 (
  id, lambda_arr
) VALUES (
  1, MAP("first name":= 15, "middle":= 25, "last name":= 0, "alt name":= 33)
);
```

Query the output.
```sql
SELECT * FROM output AS final_output;
```

> You should see something similar to:

Reviewer: ```suggestion
Your output should resemble:
```

```sql
+------------------------------+-----------------------------------------------+
|id                            |final_output                                   |
+------------------------------+-----------------------------------------------+
|1                             |{first name: 15, alt name: 33}                 |  
```

## Advanced lambda use cases

> Create a stream with a column type `MAP<STRING, ARRAY<DECIMAL(2,3)>` and apply the `transform` 

Reviewer: ```suggestion
the following example creates a stream with a column type `MAP<STRING, ARRAY<DECIMAL(2,3)>` and applies the `transform` 
```

lambda invocation function with a nested `transform` lambda invocation function.
```sql
CREATE STREAM stream1 (
  id INT,
  lambda_map MAP<STRING, ARRAY<DECIMAL(2,3)>>
) WITH (
  kafka_topic = 'stream1',
  partitions = 1,
  value_format = 'avro'
);

CREATE STREAM output AS
  SELECT id, 
  TRANSFORM(lambda_map, (k, v) => concat(k, '_new')  (k, v) => transform(v, x => round(x))) 
  FROM stream1
  EMIT CHANGES;
```
Insert some values into `stream1`.
```sql
INSERT INTO stream1 (
  id, lambda_arr
) VALUES (
  1, MAP("Mary":= ARRAY[1.23, 3.65, 8.45], "Jose":= ARRAY[5.23, 1.65]})
);
```

Query the output.
```sql
SELECT * FROM output AS final_output;
```

> You should see something similar to:

Reviewer: ```suggestion
Your output should resemble:
```

# Use lambda functions

## Context

There doesn't exist a built-in UDF that suits your needs and you're unable to implement and deploy a UDF.
To get around this limitation, you use ksqlDB implemented lambda functions.

## Syntax

The arguments for the lambda function are separated from the body of the lambda with a `=>`.

When there are 2 or more arguments the arguments must be enclosed with parentheses. Parentheses are optional for lambda functions with 1 argument.

Currently, ksqlDB only supports up to 3 arguments in a single lambda function.

```sql
x => x + 5

(x,y) => x - y

(x,y,z) => z AND x OR y
```

## Invocation UDFs

Lambda functions can only be used inside designated invocation functions. Invocations available currently are:

- [TRANSFORM](/developer-guide/ksqldb-reference/scalar-functions#TRANSFORM)
- [REDUCE](/developer-guide/ksqldb-reference/scalar-functions#REDUCE)
- [FILTER](/developer-guide/ksqldb-reference/scalar-functions#FILTER)

> ## In action

Reviewer 2: Mind moving this section just under "context" to match the others? We put that high up so people can just reference the code quickly.
